DMCompiler/DM/Visitors/DMProcBuilder.cs

```diff
@@ -51,33 +61,36 @@ class DMProcBuilder {
         }
 
         public void ProcessBlockInner(DMASTProcBlockInner block) {
-            // TODO ProcessStatementSet() needs to be before any loops but this is nasty
-            foreach (var stmt in block.Statements) {
-                if (stmt is DMASTProcStatementSet set) {
-                    try {
+            foreach (var stmt in block.SetStatements) { // Done first because all set statements are "hoisted" -- evaluated before any code in the block is run
+                Location loc = stmt.Location;
+                try {
+                    if (stmt is DMASTProcStatementSet set) {
```

Comment:
This should just do `ProcessStatement(stmt);`, that can handle the aggregate for you. You'll need to add `case DMASTProcStatementSet` and `DMASTAggregate<DMASTProcStatementSet>` in there.